src/TerminalFighter/Settings/Settings.cpp

```diff
@@ -1,44 +1,232 @@
+
+    YAML::Node node = load_node(section, keys);
+
+    if (!node.IsScalar()) {
+        LOG(ERROR) << "SettingsGroup " << vec_to_str(keys)
+                   << " from section " << static_cast<char>(section)
+                   << " is not boolean";
+        return false;
+    }
+
+    try {
+        value = node.as<bool>();
+    } catch (YAML::InvalidNode e) {
+        LOG(ERROR) << "Could not load setting " << vec_to_str(keys)
+                   << " from section " << static_cast<char>(section)
+                   << ", " << e.msg;
+        return false;
+    } catch (YAML::RepresentationException e) {
+        LOG(ERROR) << "Couldn't load " << vec_to_str(keys) << " as bool. " << e.msg;
+        return false;
+    }
+
+    return true;
+}
+
+bool Settings::load_str_map(
+    SettingsSection section,
+    std::vector<std::string> keys,
+    std::map<std::string, std::string>& value) const {
+
+    YAML::Node node = load_node(section, keys);
+
+    if (!node.IsMap()) {
+        LOG(ERROR) << "SettingsGroup " << vec_to_str(keys)
+                   << " from section " << static_cast<char>(section)
+                   << " is not a map";
+        return false;
+    }
+
+    value.clear();
+
+    for (YAML::const_iterator it = node.begin(); it != node.end(); ++it) {
+        try {
+            value[it->first.as<std::string>()] = it->second.as<std::string>();
+        } catch (YAML::InvalidNode e) {
+            LOG(ERROR) << "Could not load setting " << vec_to_str(keys)
+                       << " from section " << static_cast<char>(section)
+                       << ", " << e.msg;
+            return false;
+        } catch (YAML::RepresentationException e) {
+            LOG(ERROR) << "Couldn't load " << vec_to_str(keys) << " as map<str,str>. " << e.msg;
+            return false;
+        }
+    }
+
+    return true;
 }
+
+/* PRIVATE FUNCTIONS */
+
+int Settings::reload_setting(SettingsSection section) {
+    SettingsGroup* setting = section_to_group(section);
+
+    try {
+        setting->node_ = YAML::LoadFile(setting->file_);
+    } catch (YAML::BadFile e) {
+        LOG(ERROR) << "Failed to load file: " << setting->file_ << ", " << e.msg;
+        return 1;
+    }
+
+    LOG(INFO) << "Successfully loaded " << setting->file_;
+    return 0;
+}
+
+const Settings::SettingsGroup* Settings::section_to_group(SettingsSection section) const {
+    switch (section) {
+        case SettingsSection::VIDEO_SETTINGS:
+            return &video_settings_;
+            break;
+
+        case SettingsSection::ASSET_PATHS:
+            return &asset_paths_;
+            break;
+    }
+
+    LOG(FATAL) << "Attempted to load unknown settings section" << static_cast<char>(section);
+
+    // should never get here
+    return &video_settings_;
+}
+
+Settings::SettingsGroup* Settings::section_to_group(SettingsSection section) {
+    return const_cast<Settings::SettingsGroup*>(static_cast<const Settings*>(this)->section_to_group(section));
```

Comment:
can you explain to me what is going on here lol...

Reply:
`section_to_group` is overloaded:

```
    const SettingsGroup* section_to_group(SettingsSection section) const;
    SettingsGroup* section_to_group(SettingsSection section);
```
So default logic is to return the const version, that section of code does some nice magic to return it as non const. So now `Settings` has a const and non-const version of `section_to_group`.